### PR TITLE
Add metrics for certificate expiry

### PR DESCRIPTION
**Pull request by mitchcail**
Background:
The purpose of this PR is to introduce a task that grabs all non-expired and non-revoked certificates in lemur and emits a metric for the number of days until each certs expiration. This will give users the ability to fine tune their notifications and alerts around expiring certificates. For a bit more context see issue #3943.

### CODE REVIEW

File: lemur/common/celery.py [hunk 1]
```diff
@@ -1047,3 +1047,37 @@ def identify_expiring_deployed_certificates():
...
+        cli_certificate.expirations_metrics()
+    except SoftTimeLimitExceeded:
+        log_data["message"] = "Time limit exceeded."
+        current_app.logger.error(log_data)
+        capture_exception()
+        metrics.send("celery.timeout", "counter", 1, metric_tags={"function": function})
+        return
+
+    metrics.send(f"{function}.success", "counter", 1)
+    return log_data
```
Comment (bobmshannon) [inline lemur/common/celery.py:1083]: nit: newline

File: lemur/common/celery.py [hunk 2]
```diff
@@ -1047,3 +1047,37 @@ def identify_expiring_deployed_certificates():
...
+        "message": "sending metrics for expiring certificates",
+        "task_id": task_id,
+    }
+
+    if task_id and is_task_active(function, task_id, None):
+        log_data["message"] = "Skipping task: Task is already active"
+        current_app.logger.debug(log_data)
+        return
+
+    current_app.logger.debug(log_data)
```
Comment (bobmshannon) [inline lemur/common/celery.py:1071]: nit: remove unused debug logging

File: lemur/certificates/service.py [hunk 3]
```diff
@@ -1274,3 +1274,44 @@ def allowed_issuance_for_domain(common_name, extensions):
...
+
+    for certificate in certificates:
+        try:
+            days_until_expiration = _get_cert_expiry_in_days(certificate.not_after)
+            has_active_endpoints = len(certificate.endpoints) > 0
+            is_replacement = len(certificate.replaces) > 0
+
+
+            metrics.send(
+                f"{certificate.name}.days_until_expiration",
```
Comment (bobmshannon) [inline lemur/certificates/service.py:1295]: nit: let's use a generic metric name here e.g. `lemur.certificate.days_until_expiry`

File: lemur/certificates/cli.py [hunk 4]
```diff
@@ -1118,3 +1119,25 @@ def identify_expiring_deployed_certificates(exclude_domains, exclude_owners, com
...
+def expirations_metrics():
+    """
+    Iterates over all certificates and emits a metric for the days remaining for a certificate to expire.
+    This is used for building custom dashboards and alerts for certificate expiry.
+    """
+    try:
+        print("Starting to publish metrics for time left until cert expirations")
+        success, failure = send_certificate_expiration_metrics()
+        print(
+            f"Finished publishing metrics for time left until cert expirations! Sent: {success} "
```
Comment (bobmshannon) [inline lemur/certificates/cli.py:1134]:
```suggestion
            f"Finished publishing metrics for time left until cert expirations! Sent: {success}"
```

File: lemur/certificates/service.py [hunk 5]
```diff
@@ -1274,3 +1274,43 @@ def allowed_issuance_for_domain(common_name, extensions):
...
         is_authorized_for_domain(common_name)
+
+
+def send_certificate_expiration_metrics():
+    """
+    Iterate over each certificate and emit a metric for how many days until expiration.
+    """
+    success = failure = 0
+    # get eligible certificates
+    certificates = get_certificates()
```
Comment (bobmshannon) [inline lemur/certificates/service.py:1285]: This seems to filter out certificates which have notifications disabled and are not expiring within 30 days. We want this to monitor all non-expired certificates so this behavior is probably not what we want here. So I'd suggest creating another function here which does something similar as  `get_certificates()` but has the required filtering constraints.

Comment (mitchcail) [reply lemur/certificates/service.py:1285]: Understood, so maybe something like,
```
certificates = database.db.session.query(Certificate).filter(
        Certificate.expired == false(),
        Certificate.revoked == false()
    )
```

**Review by bobmshannon — APPROVED** [review]

File: lemur/certificates/cli.py [hunk 6]
```diff
@@ -1118,3 +1119,25 @@
...
+def expiration_metrics():
+    """
+    Iterates over all certificates and emits a metric for the days remaining for a certificate to expire.
+    This is used for building custom dashboards and alerts for certificate expiry.
+    """
+    try:
+        print("Starting to publish metrics for time left until cert expirations")
+        success, failure = send_certificate_expiration_metrics()
+        print(
+            f"Finished publishing metrics for time left until cert expirations! Sent: {success}"
```
[the remark this answers is not in the document]
Comment (mitchcail) [reply lemur/certificates/cli.py:1134]: Please let me know if anything needs to be actioned around this. I was trying to follow a similar pattern to this [method](https://github.com/Netflix/lemur/blob/master/lemur/notifications/cli.py#L52-L59).

File: lemur/certificates/service.py [hunk 7]
```diff
@@ -1274,3 +1274,46 @@ def allowed_issuance_for_domain(common_name, extensions):
...
+def send_certificate_expiration_metrics():
+    """
+    Iterate over each certificate and emit a metric for how many days until expiration.
+    """
+    success = failure = 0
+
+    certificates = database.db.session.query(Certificate).filter(
+        Certificate.expired == false(),
+        Certificate.revoked == false()
+    )
```
Comment (jtschladen) [inline lemur/certificates/service.py:1288]: Would it make sense to include a (perhaps optional) parameter limiting the query to a maximum expiry period (probably in days)? Iterating over all valid certs could potentially be a very large data set in a large Lemur deployment, and I'm guessing this metric might be most valuable for understanding certs expiring soon -- or is getting visibility into the full set of certs more the goal?

[the remark this answers is not in the document]
Comment (mitchcail) [reply lemur/certificates/service.py]: I updated this query to filter out already rotated certs in commit [c415dca](https://github.com/Netflix/lemur/pull/4084/commits/c415dcaf746e76365592de9446a8e3e768b5234c)

Comment (mitchcail) [reply lemur/certificates/service.py]: Ability to pass in an expiry window was added as part of commit [e31ae99](https://github.com/Netflix/lemur/pull/4084/commits/e31ae99463418aab1a7ddbf0c76308e2d7129dac). The default of None will query for all active certs, while passing in a value such as 90 will only query for active certs expiring over the next 90 days. As part of this commit I moved the query logic to a separate method to make testing that logic easier.

**Review by jtschladen — APPROVED** [review]
Thanks for the contribution @mitchcail !

**Pull request merged**